include/gridtools/reduction/gpu.hpp

```diff
@@ -0,0 +1,324 @@
+
+            template <class Kernel, class F, class T>
+            __global__ void launch(Kernel kernel, F f, T const *__restrict__ in, T *__restrict__ out) {
+                kernel(f, in, out);
+            }
+
+            inline constexpr bool is_pow2(size_t x) { return (x & x - 1) == 0; }
+
+            inline constexpr size_t next_pow2(size_t x) {
+                --x;
+                x |= x >> 1;
+                x |= x >> 2;
+                x |= x >> 4;
+                x |= x >> 8;
+                x |= x >> 16;
+                x |= x >> 32;
+                return ++x;
+            }
+
+            inline constexpr size_t int_log2(size_t val) {
+                assert(val);
+                size_t ret = 0;
+                for (; val > 1; val /= 2)
+                    ret++;
+                return ret;
+            }
+
+            // TODO(anstaf): benchmark that
+            using cpu_final_threshold_t = integral_constant<size_t, 8>;
+
+            static_assert(cpu_final_threshold_t::value > 1, GT_INTERNAL_ERROR);
+            static_assert(is_pow2(cpu_final_threshold_t::value), GT_INTERNAL_ERROR);
+
+            inline auto get_device_properties() {
+                int device;
+                GT_CUDA_CHECK(cudaGetDevice(&device));
+                cudaDeviceProp res;
+                GT_CUDA_CHECK(cudaGetDeviceProperties(&res, device));
+                return res;
+            }
+
+            inline size_t max_threads() {
+                static size_t res = get_device_properties().maxThreadsPerBlock;
+                return res;
+            }
+
+            inline size_t get_threads(size_t n) {
+                assert(is_pow2(max_threads()));
+                assert(n % 2 == 0);
+                return gcd(n / 2, max_threads());
+            }
+
+            template <class T>
+            double kahan_sum(T const *data, size_t n) {
+                assert(n);
+                double sum = data[0];
+                double c = 0;
+                for (int i = 1; i < n; i++) {
+                    double y = data[i] - c;
+                    double t = sum + y;
+                    c = (t - sum) - y;
+                    sum = t;
+                }
+                return sum;
+            }
+
+            template <class F, class T>
+            T reduce_cpu(F f, T const *buff, size_t n) {
+                assert(n);
+                T res = buff[0];
+                for (size_t i = 1; i != n; i++)
+                    res = f(res, buff[i]);
+                return res;
+            }
+
+            inline auto reduce_cpu(plus, float const *buff, size_t n) { return kahan_sum(buff, n); }
+
+            inline auto reduce_cpu(plus, double const *buff, size_t n) { return kahan_sum(buff, n); }
+
+            template <class F, class T>
+            void reduce_step(F f, size_t threads, size_t blocks, T const *__restrict__ in, T *__restrict__ out) {
+                constexpr size_t max_threads = 1024;
```

Comment:
Not sure if it is a good idea to assume that the max thread count is always 1024 or if it might be better to query `deviceProperties`.